CompiletoDalvik doesn't like non-ASCII characters on Windows

See also: https://developercommunity.visualstudio.com/content/problem/437807/v-isual-studio-2019-preview-20.html
Work item: http://work.devdiv.io/784551

### Steps to Reproduce

1. Create a new project containing non-ASCII characters, *on Windows*.
2. Install the project to an Android target

-- or --

1. Download [Scratch.Ümläüts-Win.zip](https://github.com/xamarin/xamarin-android/files/2823843/Scratch.Umlauts-Win.zip)
2. Extract `Scratch.Ümläüts-Win.zip` using Windows Explorer
3. Build `Scratch.Ümläüts.csproj`:

        cd Scratch.Ümläüts\Scratch.Ümläüts
        msbuild /v:diag Scratch.Ümläüts.csproj /t:SignAndroidPackage > b.txt

(and because *file encodings within `.zip` files are still a problem in **2019***, here's a Unix-formatted `.zip` file with the same contents, but extractable from Unix/WSL `unzip` but *not* Windows Explorer: [Scratch.Ümläüts-unix.zip](https://github.com/xamarin/xamarin-android/files/2823842/Scratch.Umlauts-unix.zip).)

### Expected Behavior

It installs!

### Actual Behavior

It doesn't. :-(

### Version Information

Visual Studio 2019 Preview 2.

### Log File

```
1>Task "CompileToDalvik" (TaskId:245)
1>  Task Parameter:DxJarPath=C:\Program Files (x86)\Android\android-sdk\build-tools\25.0.3\\lib\dx.jar (TaskId:245)
1>  Task Parameter:DxExtraArguments=--dex --no-strict (TaskId:245)
1>  Task Parameter:JavaToolPath=C:\Program Files\Android\jdk\microsoft_dist_openjdk_1.8.0.9\bin (TaskId:245)
1>  Task Parameter:JavaMaximumHeapSize=1G (TaskId:245)
1>  Task Parameter:ClassesOutputDirectory=obj\Debug\90\android\bin\classes (TaskId:245)
1>  Task Parameter:MultiDexEnabled=False (TaskId:245)
1>  Task Parameter:MultiDexMainDexListFile=obj\Debug\90\multidex.keep (TaskId:245)
1>  Task Parameter:JavaLibrariesToCompile=C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\Common7\IDE\ReferenceAssemblies\Microsoft\Framework\MonoAndroid\v9.0\mono.android.jar (TaskId:245)
1>    processing ClassesOutputDirectory... (TaskId:245)
1>      C:\Temp\Scratch.Ümläüts\Scratch.Ümläüts\obj\Debug\90\android\bin\classes.zip (TaskId:245)
1>      C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\Common7\IDE\ReferenceAssemblies\Microsoft\Framework\MonoAndroid\v9.0\mono.android.jar (TaskId:245)
1>  C:\Program Files\Android\jdk\microsoft_dist_openjdk_1.8.0.9\bin\java.exe -Xmx1G -jar "C:\Program Files (x86)\Android\android-sdk\build-tools\25.0.3\\lib\dx.jar" --dex --no-strict --input-list="C:\Users\Windows User\AppData\Local\Temp\tmp6A5B.tmp" --output obj\Debug\90\android\bin  (TaskId:245)
1>   (TaskId:245)
1>  UNEXPECTED TOP-LEVEL EXCEPTION: (TaskId:245)
1>  java.io.FileNotFoundException: C:\Temp\Scratch.├£ml├ñ├╝ts\Scratch.├£ml├ñ├╝ts\obj\Debug\90\android\bin\classes.zip (The system cannot find the path specified) (TaskId:245)
1>  	at java.util.zip.ZipFile.open(Native Method) (TaskId:245)
1>  	at java.util.zip.ZipFile.<init>(ZipFile.java:219) (TaskId:245)
1>  	at java.util.zip.ZipFile.<init>(ZipFile.java:149) (TaskId:245)
1>  	at java.util.zip.ZipFile.<init>(ZipFile.java:163) (TaskId:245)
1>  	at com.android.dx.cf.direct.ClassPathOpener.processArchive(ClassPathOpener.java:244) (TaskId:245)
1>  	at com.android.dx.cf.direct.ClassPathOpener.processOne(ClassPathOpener.java:166) (TaskId:245)
1>  	at com.android.dx.cf.direct.ClassPathOpener.process(ClassPathOpener.java:144) (TaskId:245)
1>  	at com.android.dx.command.dexer.Main.processOne(Main.java:677) (TaskId:245)
1>  	at com.android.dx.command.dexer.Main.processAllFiles(Main.java:574) (TaskId:245)
1>  	at com.android.dx.command.dexer.Main.runMonoDex(Main.java:311) (TaskId:245)
1>  	at com.android.dx.command.dexer.Main.run(Main.java:277) (TaskId:245)
1>  	at com.android.dx.command.dexer.Main.main(Main.java:245) (TaskId:245)
1>  	at com.android.dx.command.Main.main(Main.java:106) (TaskId:245)
1>  1 error; aborting (TaskId:245)
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(2668,3): error MSB6006: "java.exe" exited with code 1.
1>Done executing task "CompileToDalvik" -- FAILED. (TaskId:245)
1>Done building target "_CompileToDalvikWithDx" in project "Scratch.Ümläüts.csproj" -- FAILED.: (TargetId:150)
```

### Additional Information

This builds as expected/desired on d15-9/Xamarin.Android 9.1.0:

```
Task "CompileToDalvik" (TaskId:211)
  Task Parameter:DxJarPath=C:\Program Files (x86)\Android\android-sdk\build-tools\25.0.3\\lib\dx.jar (TaskId:211)
  Task Parameter:DxExtraArguments=--dex --no-strict (TaskId:211)
  Task Parameter:JavaToolPath=C:\Program Files\Java\jdk1.8.0_152\\bin (TaskId:211)
  Task Parameter:JavaMaximumHeapSize=1G (TaskId:211)
  Task Parameter:ClassesOutputDirectory=obj\Debug\81\android\bin\classes (TaskId:211)
  Task Parameter:MultiDexEnabled=False (TaskId:211)
  Task Parameter:MultiDexMainDexListFile=obj\Debug\81\multidex.keep (TaskId:211)
  Task Parameter:JavaLibrariesToCompile=C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Common7\IDE\ReferenceAssemblies\Microsoft\Framework\MonoAndroid\v8.1\mono.android.jar (TaskId:211)
  CompileToDalvik (TaskId:211)
    JavaOptions:  (TaskId:211)
    JavaMaximumHeapSize: 1G (TaskId:211)
    ClassesOutputDirectory: obj\Debug\81\android\bin\classes (TaskId:211)
    JavaToolPath: C:\Program Files\Java\jdk1.8.0_152\\bin (TaskId:211)
    DxJarPath: C:\Program Files (x86)\Android\android-sdk\build-tools\25.0.3\\lib\dx.jar (TaskId:211)
    ToolExe: java.exe (TaskId:211)
    ToolPath:  (TaskId:211)
    UseDx: False (TaskId:211)
    DxExtraArguments: --dex --no-strict (TaskId:211)
    MultiDexEnabled: False (TaskId:211)
    MultiDexMainDexListFile: obj\Debug\81\multidex.keep (TaskId:211)
    JavaLibrariesToCompile: (TaskId:211)
      C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Common7\IDE\ReferenceAssemblies\Microsoft\Framework\MonoAndroid\v8.1\mono.android.jar (TaskId:211)
    AlternativeJarFiles: (TaskId:211)
    processing ClassesOutputDirectory... (TaskId:211)
  C:\Program Files\Java\jdk1.8.0_152\\bin\java.exe -Xmx1G -jar "C:\Program Files (x86)\Android\android-sdk\build-tools\25.0.3\\lib\dx.jar" --dex --no-strict --output obj\Debug\81\android\bin "C:\Users\jopryo\Downloads\Scratch.�ml��ts\Scratch.�ml��ts\obj\Debug\81\android\bin\classes.zip" "C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Common7\IDE\ReferenceAssemblies\Microsoft\Framework\MonoAndroid\v8.1\mono.android.jar"  (TaskId:211)
    DexOutputs:  (TaskId:211)
      obj\Debug\81\android\bin\classes.dex (TaskId:211)
Done executing task "CompileToDalvik". (TaskId:211)
```

(Note "garbage characters" in the log file, because *file encoding is hard, y'all*!)

This was probably broken by commit https://github.com/xamarin/xamarin-android/commit/8571a7f696e864a791f726dcd2fb856145a53ee1, which introduced use of `--input-list` via a "temporary file" `inputListFile` which contains the `.jar` files to process instead of providing them on the command-line.

My *suspicion* is that the fix will need to take a page out of https://github.com/xamarin/monodroid/commit/1c5650552d31e85df5a58cefbbd448691a186f04:

1. Ensure that `inputListFile` is UTF-8 (already done)
2. Ensure that filenames within `inputListFile` use the correct normalization form (*perhaps* already done?)
3. Launch the JVM with `-Dfile.encoding=UTF8` so that it (hopefully) reads files as UTF-8 and *not "**ANSI**"*.

--- a/src/Xamarin.Android.Build.Tasks/Tasks/AndroidUpdateResDir.cs
+++ b/src/Xamarin.Android.Build.Tasks/Tasks/AndroidUpdateResDir.cs
@@ -25,6 +25,7 @@
 // THE SOFTWARE.
 
 using System;
+using System.Text;
 using Microsoft.Build.Utilities;
 using Microsoft.Build.Framework;
 using System.IO;
@@ -96,8 +97,8 @@ namespace Xamarin.Android.Tasks
 					if (string.IsNullOrEmpty (rel)) {
 						rel = item.GetMetadata ("Identity");
 						if (!string.IsNullOrEmpty (ProjectDir)) {
-							var fullRelPath = Path.GetFullPath (rel);
-							var fullProjectPath = Path.GetFullPath (ProjectDir);
+							var fullRelPath = Path.GetFullPath (rel).Normalize (NormalizationForm.FormC);
+							var fullProjectPath = Path.GetFullPath (ProjectDir).Normalize (NormalizationForm.FormC);
 							if (fullRelPath.StartsWith (fullProjectPath)) {
 								rel = fullRelPath.Replace (fullProjectPath, string.Empty);
 							}

--- a/src/Xamarin.Android.Build.Tasks/Tasks/CompileToDalvik.cs
+++ b/src/Xamarin.Android.Build.Tasks/Tasks/CompileToDalvik.cs
@@ -82,11 +82,18 @@ namespace Xamarin.Android.Tasks
 					cmd.AppendSwitch (JavaOptions);		
 				}
 
+				cmd.AppendSwitchIfNotNull ("-Dfile.encoding=", "UTF8");
 				// Add the specific -XmxN to override the default heap size for the JVM
 				// N can be in the form of Nm or NGB (e.g 100m or 1GB ) 
 				cmd.AppendSwitchIfNotNull("-Xmx", JavaMaximumHeapSize);
 
 				cmd.AppendSwitchIfNotNull ("-jar ", Path.Combine (DxJarPath));
+			} else {
+				// To pass additional java parameters to `dx` you must 
+				// provide the parameter without the leading `-` 
+				// the dx tool will add that in after stripping off
+				// the `-J`
+				cmd.AppendSwitchIfNotNull ("-JDfile.encoding=", "UTF8");
 			}
 
 			cmd.AppendSwitch (DxExtraArguments);

--- a/src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/BuildTest.cs
+++ b/src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/BuildTest.cs
@@ -2332,7 +2332,7 @@ public class Test
 		}
 
 		[Test]
-		public void BuildApplicationWithSpacesInPath ([Values (true, false)] bool enableMultiDex, [Values ("dx", "d8")] string dexTool, [Values ("", "proguard", "r8")] string linkTool)
+		public void BuildApplicationWithSpacesInPathAndÜmläüts ([Values (true, false)] bool enableMultiDex, [Values ("dx", "d8")] string dexTool, [Values ("", "proguard", "r8")] string linkTool)
 		{
 			var proj = new XamarinAndroidApplicationProject () {
 				IsRelease = true,
@@ -2368,7 +2368,7 @@ AAMMAAABzYW1wbGUvSGVsbG8uY2xhc3NQSwUGAAAAAAMAAwC9AAAA1gEAAAAA") });
 </Project>
 ",
 			});
-			using (var b = CreateApkBuilder (Path.Combine ("temp", $"BuildReleaseAppWithA InIt({enableMultiDex}{dexTool}{linkTool})"))) {
+			using (var b = CreateApkBuilder (Path.Combine ("temp", $"BuildReleaseAppWithA InItAndÜmläüts({enableMultiDex}{dexTool}{linkTool})"))) {
 				Assert.IsTrue (b.Build (proj), "Build should have succeeded.");
 				Assert.IsFalse (b.LastBuildOutput.ContainsText ("Duplicate zip entry"), "Should not get warning about [META-INF/MANIFEST.MF]");
 

